File: poc4/core/eval.js
"use strict";

const fs = require("fs");
const path = require("path");
const parser = require("./parser");
const kernel = require("./kernel");

const ROOT = path.dirname(path.dirname(path.resolve(__filename)));


// === Env ===

class Env {
    constructor(d) { this.d = d || {}; }
    get(name) {
        if (!(name in this.d)) throw new Error(`undefined: ${name}`);
        return this.d[name];
    }
    snapshot() { return new Env({ ...this.d }); }
}


// === Values ===

class Fn {
    constructor(params, bodyFn, env) {
        this.params = params;
        this.bodyFn = bodyFn;
        this.env = env;
    }
    toString() { return `\\${this.params.join(" ")}: ...`; }
}


// === Errors ===

class EvalError extends Error {
    constructor(msg, node) {
        super(msg);
        this.msg = msg;
        this.node = node;
        this.trace = [];
    }
    addFrame(node) { this.trace.push(node); }
    toString() {
        const lines = [];
        if (this.node)
            lines.push(`${this.node.file}:${this.node.line}:${this.node.col}: ${this.msg}`);
        else
            lines.push(this.msg);
        for (const n of this.trace)
            lines.push(`  at ${n.file}:${n.line}:${n.col}`);
        return lines.join("\n");
    }
}


// === Compiler: AST -> (env) => value ===

function compile(node) {
    if (node instanceof parser.Lit) {
        const v = node.value;
        return (_env) => v;
    }
    if (node instanceof parser.Ref) {
        const name = node.name;
        return (env) => {
            if (!(name in env.d)) throw new EvalError(`undefined: ${name}`, node);
            return env.d[name];
        };
    }
    if (node instanceof parser.Fn) {
        const params = node.params;
        const bodyFn = compile(node.body);
        return (env) => new Fn(params, bodyFn, env.snapshot());
    }
    if (node instanceof parser.Block) {
        const binds = node.bindings.map(b => ({ name: b.name, fn: compile(b.expr) }));
        const exprFn = compile(node.expr);
        return (env) => {
            env = env.snapshot();
            for (const b of binds) env.d[b.name] = b.fn(env);
            return exprFn(env);
        };
    }
    if (node instanceof parser.If) {
        const condFn = compile(node.cond);
        const thenFn = compile(node.then);
        const elseFn = compile(node.else_);
        return (env) => {
            const c = condFn(env);
            if (c === true) return thenFn(env);
            if (c === false) return elseFn(env);
            throw new EvalError("condition must be true or false", node);
        };
    }
    if (node instanceof parser.Call) {
        const calleeFn = compile(node.callee);
        const argFns = node.args.map(a => compile(a));
        return (env) => {
            const callee = calleeFn(env);
            const args = argFns.map(fn => fn(env));
            try { return call(callee, args, node); }
            catch (e) {
                if (e instanceof EvalError) { e.addFrame(node); throw e; }
                throw new EvalError(e.message, node);
            }
        };
    }
    if (node instanceof parser.List) {
        const fns = node.elems.map(e => compile(e));
        return (env) => fns.map(f => f(env));
    }
    throw new EvalError(`unknown: ${node.constructor.name}`, node);
}


function exec(callee, args) {
    if (callee instanceof Fn) {
        if (args.length !== callee.params.length)
            throw new EvalError(`arity: expected ${callee.params.length} args, got ${args.length}`);
        const env = callee.env.snapshot();
        for (let i = 0; i < callee.params.length; i++)
            env.d[callee.params[i]] = args[i];
        return callee.bodyFn(env);
    }
    if (callee._arity !== undefined && args.length !== callee._arity)
        throw new EvalError(`arity: expected ${callee._arity} args, got ${args.length}`);
    return callee(...args);
}

const _R = Symbol();

function cacheGet(m, args) {
    for (const a of args) { m = m.get(a); if (!m) return undefined; }
    return m.get(_R);
}

function cacheSet(m, args, val) {
    for (const a of args) { let n = m.get(a); if (!n) { n = new Map(); m.set(a, n); } m = n; }
    m.set(_R, val);
}

function call(callee, args, node) {
    if (!callee._c) { callee._c = new Map(); callee._n = 0; }
    const hit = cacheGet(callee._c, args);
    if (hit !== undefined) return hit;
    if (callee._n > 1024) { callee._c.clear(); callee._n = 0; }
    const result = exec(callee, args);
    cacheSet(callee._c, args, result);
    callee._n++;
    return result;
}


// === Formula bridge ===

const _buildCache = new WeakMap();
function build(f) {
    if (typeof f === "string") return f;
    const cached = _buildCache.get(f);
    if (cached) return cached;
    const tag = f[0];
    let result;
    if (tag === "mem") result = new kernel.Mem(build(f[1]), build(f[2]));
    else if (tag === "neg") result = new kernel.Neg(build(f[1]));
    else if (tag === "implies") result = new kernel.Implies(build(f[1]), build(f[2]));
    else if (tag === "forall") result = new kernel.Forall(f[1], build(f[2]));
    else throw new Error(`build: unknown tag ${tag}`);
    _buildCache.set(f, result);
    return result;
}

function doProof(left, right, rule, premises, principal, term) {
    try {
        const seq = new kernel.Sequent(left.map(build), right.map(build));
        return [true, kernel.proof(seq, rule, premises, build(principal), term)];
    } catch (e) {
        return [false, e.message];
    }
}

function doQed(p, expected, system) {
    try {
        kernel.qed(p, build(expected), system);
        return [true, null];
    } catch (e) {
        return [false, e.message];
    }
}

function fail(msg) { throw new EvalError(msg); }

// === Builtins ===

function builtin(arity, fn) {
    fn._arity = arity;
    return fn;
}

function makeGlobal() {
    return new Env({
        true: true, false: false, none: null,
        is_none: builtin(1, (a) => a === null),
        add: builtin(2, (a, b) => {
            if (a === null || b === null) throw new EvalError("add: null argument");
            return Array.isArray(a) ? [...a, ...b] : a + b;
        }),
        sub: builtin(2, (a, b) => a - b),
        mul: builtin(2, (a, b) => a * b),
        eq: builtin(2, (a, b) => {
            if (typeof a !== "string" && typeof a !== "number" && typeof a !== "boolean")
                throw new EvalError(`eq: unsupported type ${typeof a}`);
            if (typeof a !== typeof b)
                throw new EvalError(`eq: type mismatch ${typeof a} vs ${typeof b}`);
            return a === b;
        }),
        str: builtin(1, (a) => String(a)),
        not: builtin(1, (a) => !a),
        head: builtin(1, (a) => a[0]),
        tail: builtin(1, (a) => a.slice(1)),
        nth: builtin(2, (a, n) => a[n]),
        len: builtin(1, (a) => a.length),
        tap: builtin(1, (a) => { console.log(a); return a; }),
        _do_proof: builtin(6, doProof),
        _do_qed: builtin(3, doQed),
        _fail: builtin(1, fail),
    });
}


// === File loading ===

const loaded = {};

function loadFile(filepath) {
    filepath = path.resolve(filepath);
    if (filepath in loaded) return loaded[filepath];
    loaded[filepath] = {};

    const source = fs.readFileSync(filepath, "utf-8");
    const nodes = parser.parse(source, filepath);
    const env = makeGlobal();

    for (const node of nodes) {
        if (node instanceof parser.Import) {
            loadImport(node, env);
        } else if (node instanceof parser.Bind) {
            const fn = compile(node.expr);
            const val = fn(env);
            env.d[node.name] = val;
            loaded[filepath][node.name] = val;
        }
    }

    return loaded[filepath];
}

function loadImport(node, env) {
    const parts = node.module.split(".");
    const filepath = path.join(ROOT, ...parts) + ".min";
    if (!fs.existsSync(filepath))
        throw new EvalError(`module not found: ${node.module} (${filepath})`);
    const imported = loadFile(filepath);
    for (const { name, alias } of node.names) {
        if (name.startsWith("_"))
            throw new EvalError(`cannot import private name "${name}" from ${node.module}`);
        if (!(name in imported))
            throw new EvalError(`name "${name}" not found in ${node.module}`);
        env.d[alias] = imported[name];
    }
}

function run(filepath) {
    try {
        loadFile(filepath);
        return true;
    } catch (e) {
        if (e instanceof EvalError) {
            process.stderr.write(e.toString() + "\n");
            return false;
        }
        throw e;
    }
}


// === CLI ===

if (require.main === module) {
    if (process.argv.length > 2) {
        const ok = run(process.argv[2]);
        process.exit(ok ? 0 : 1);
    }

    const src = `
$double \\n: add(n, n)
$r1 tap(double(3))
$r2 tap(add("double(1+2) = ", str(double(add(1, 2)))))

$id \\x: x
$r3 tap(id(42))

$pair \\a b: [a, b]
$r4 tap(pair(1, 2))

$compose \\f g: \\x: f(g(x))
$quadruple compose(double, double)
$r5 tap(quadruple(3))

$fact \\self n: ?(eq(n, 0), 1, mul(n, self(self, sub(n, 1))))
$r6 tap(fact(fact, 5))
`;
    const env = makeGlobal();
    const nodes = parser.parse(src, "<test>");
    for (const node of nodes) {
        if (node instanceof parser.Bind) {
            const fn = compile(node.expr);
            env.d[node.name] = fn(env);
        }
    }
}

module.exports = { Env, Fn, EvalError, compile, call, loadFile, run };

File: poc3/core/eval.py
"""Pure eval. Closures, immutable env, thin Python wrapper."""

import os, sys, parser, kernel

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


# === Env (immutable extend) ===

class Env:
    __slots__ = ('d',)
    def __init__(self, d=None):
        self.d = d if d is not None else {}

    def get(self, name):
        return self.d[name]

    def snapshot(self):
        return Env(self.d.copy())


# === Values (Python objects) ===

class Fn:
    def __init__(self, params, body, env):
        self.params = params
        self.body = body
        self.env = env

    def __repr__(self):
        return f'\\{" ".join(self.params)}: ...'


# === Eval ===

class EvalError(Exception):
    def __init__(self, msg, node=None):
        self.msg = msg
        self.node = node
        self.trace = []
    def add_frame(self, node):
        self.trace.append(node)
    def __str__(self):
        lines = []
        if self.node:
            lines.append(f'{self.node.file}:{self.node.line}:{self.node.col}: {self.msg}')
        else:
            lines.append(self.msg)
        for n in self.trace:
            lines.append(f'  at {n.file}:{n.line}:{n.col}')
        return '\n'.join(lines)


def _eval_lit(node, env):
    return node.value

def _eval_ref(node, env):
    try:
        return env.get(node.name)
    except KeyError:
        raise EvalError(f'undefined: {node.name}', node)

def _eval_fn(node, env):
    return Fn(node.params, node.body, env.snapshot())

def _eval_block(node, env):
    env = env.snapshot()
    for binding in node.bindings:
        val = evaluate(binding.expr, env)
        env.d[binding.name] = val
    return evaluate(node.expr, env)

def _eval_if(node, env):
    cond = evaluate(node.cond, env)
    if cond is True:
        return evaluate(node.then, env)
    if cond is False:
        return evaluate(node.else_, env)
    raise EvalError('condition must be True or False', node)

def _eval_call(node, env):
    callee = evaluate(node.callee, env)
    args = [evaluate(a, env) for a in node.args]
    try:
        return call(callee, args, node)
    except EvalError as e:
        e.add_frame(node)
        raise

def _eval_list(node, env):
    return [evaluate(e, env) for e in node.elems]

_DISPATCH = {
    parser.Lit: _eval_lit,
    parser.Ref: _eval_ref,
    parser.Fn: _eval_fn,
    parser.Block: _eval_block,
    parser.If: _eval_if,
    parser.Call: _eval_call,
    parser.List: _eval_list,
}

def evaluate(node, env):
    f = _DISPATCH.get(type(node))
    if f:
        return f(node, env)
    raise EvalError(f'unknown: {type(node).__name__}', node)


def call(callee, args, node=None):
    if isinstance(callee, Fn):
        env = callee.env.snapshot()
        for i, p in enumerate(callee.params):
            env.d[p] = args[i] if i < len(args) else None
        return evaluate(callee.body, env)
    if callable(callee):
        return callee(*args)
    raise EvalError(f'not callable: {callee}', node)


# === Formula bridge ===

def build(f):
    if isinstance(f, str):
        return f
    tag = f[0]
    if tag == 'mem': return kernel.Mem(build(f[1]), build(f[2]))
    if tag == 'neg': return kernel.Neg(build(f[1]))
    if tag == 'implies': return kernel.Implies(build(f[1]), build(f[2]))
    if tag == 'forall': return kernel.Forall(f[1], build(f[2]))
    raise ValueError(f'build: unknown tag {tag}')

def _do_proof(left, right, rule, premises, principal, term=None):
    try:
        seq = kernel.Sequent(
            [build(f) for f in left],
            [build(f) for f in right])
        return [True, kernel.proof(seq, rule, premises, build(principal), term)]
    except (ValueError, TypeError) as e:
        return [False, str(e)]

def _do_qed(p, expected, system):
    try:
        kernel.qed(p, build(expected), system)
        return [True, None]
    except (ValueError, TypeError) as e:
        return [False, str(e)]

def _fail(msg):
    raise EvalError(msg)


# === Builtins ===

def _global():
    return Env({
        'True': True, 'False': False, 'None': None,
        'add': lambda a, b: a + b, 'sub': lambda a, b: a - b,
        'mul': lambda a, b: a * b, 'eq': lambda a, b: a == b,
        'not': lambda a: not a,
        'head': lambda a: a[0], 'tail': lambda a: a[1:],
        'len': lambda a: len(a), 'print': lambda a: print(a) or a,
        '_do_proof': _do_proof, '_do_qed': _do_qed, '_fail': _fail,
    })


# === File loading ===

_loaded = {}

def load_file(filepath):
    filepath = os.path.normpath(filepath)
    if filepath in _loaded:
        return _loaded[filepath]
    _loaded[filepath] = {}

    with open(filepath) as f:
        source = f.read()
    nodes = parser.parse(source, filepath)
    env = _global()

    for node in nodes:
        if isinstance(node, parser.Import):
            _load_import(node, env)
        elif isinstance(node, parser.Bind):
            val = evaluate(node.expr, env)
            env.d[node.name] = val
            _loaded[filepath][node.name] = val

    return _loaded[filepath]


def _load_import(node, env):
    parts = node.module.split('.')
    filepath = os.path.join(ROOT, *parts) + '.min'
    if not os.path.isfile(filepath):
        return
    imported = load_file(filepath)
    for name in node.names:
        if name in imported:
            env.d[name] = imported[name]


def run(filepath):
    try:
        load_file(filepath)
    except EvalError as e:
        print(e, file=sys.stderr)
        return False
    return True


# === Self-test ===

if __name__ == '__main__':
    if len(sys.argv) > 1:
        ok = run(sys.argv[1])
        sys.exit(0 if ok else 1)

    src = r'''
$double \n: add(n, n)
$r1 print(double(3))
$r2 print(double(add(1, 2)))

$id \x: x
$r3 print(id(42))

$pair \a b: [a, b]
$r4 print(pair(1, 2))

$compose \f g: \x: f(g(x))
$quadruple compose(double, double)
$r5 print(quadruple(3))

$fact \self n: ?(eq(n, 0), 1, mul(n, self(self, sub(n, 1))))
$r6 print(fact(fact, 5))
'''
    env = _global()
    nodes = parser.parse(src, '<test>')
    for node in nodes:
        if isinstance(node, parser.Bind):
            val = evaluate(node.expr, env)
            env = env.extend(node.name, val)
